File: DDCharacterCreator/app/src/main/java/com/example/ddcharactercreator/SpellAdapter.java
package com.example.ddcharactercreator;

import android.content.Context;
import android.content.DialogInterface;
import android.content.Intent;
import android.os.Build;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;
import android.widget.TextView;
import androidx.annotation.NonNull;
import androidx.annotation.RequiresApi;
import androidx.recyclerview.widget.RecyclerView;
import java.util.ArrayList;
import java.util.List;
import android.app.AlertDialog;
import android.widget.Toast;
import static com.example.ddcharactercreator.DetailActivity.cantripCount;
import static com.example.ddcharactercreator.DetailActivity.spellCount;

public class SpellAdapter extends RecyclerView.Adapter<SpellAdapter.SpellViewHolder> {

    private Context context;
    private List<Spell> spells = new ArrayList<>();
    private Boolean isDatabaseSpell;
    private Character character = DetailActivity.character;
    private ArrayList<Spell> cantripsList = new ArrayList<>();
    private ArrayList<Spell> spellsList = new ArrayList<>();

    public SpellAdapter(@NonNull Context context, List<Spell> spells, Boolean isDatabaseSpell) {
        this.context = context;
        this.spells = spells;
        this.isDatabaseSpell = isDatabaseSpell;
    }

    @NonNull
    @Override
    public SpellAdapter.SpellViewHolder onCreateViewHolder(@NonNull ViewGroup parent, int viewType) {
        return new SpellViewHolder(LayoutInflater.from(context).inflate(R.layout.spell_list_item, parent, false));
    }

    @Override
    public void onBindViewHolder(@NonNull SpellViewHolder holder, int position) {
        for(int i=0; i<character.getSpellsKnown().size(); i++){
            Spell thisSpell = character.getSpellsKnown().get(i);
            if(thisSpell.getLevel() == 0 && !cantripsList.contains(thisSpell)){
                cantripsList.add(thisSpell);
            } else if(!cantripsList.contains(thisSpell) && !spellsList.contains(thisSpell)){
                spellsList.add(thisSpell);
            }
        }
        Spell spell = spells.get(position);
        holder.spellNameTextView.setText(spell.getSpellName());
        if(spell.getLevel() > 0){
            holder.spellLevelTextView.setText(String.valueOf(spell.getLevel()));
        } else{
            holder.spellLevelTextView.setVisibility(View.GONE);
        }
        switch(spell.getSchool()){
            case "Abjuration":
                holder.spellNameTextView.setTextColor(context.getResources().getColor(R.color.abjuration));
                break;
            case "Conjuration":
                holder.spellNameTextView.setTextColor(context.getResources().getColor(R.color.conjuration));
                break;
            case "Divination":
                holder.spellNameTextView.setTextColor(context.getResources().getColor(R.color.divination));
                break;
            case "Enchantment":
                holder.spellNameTextView.setTextColor(context.getResources().getColor(R.color.enchantment));
                break;
            case "Evocation":
                holder.spellNameTextView.setTextColor(context.getResources().getColor(R.color.evocation));
                break;
            case "Illusion":
                holder.spellNameTextView.setTextColor(context.getResources().getColor(R.color.illusion));
                break;
            case "Necromancy":
                holder.spellNameTextView.setTextColor(context.getResources().getColor(R.color.necromancy));
                break;
            case "Transmutation":
                holder.spellNameTextView.setTextColor(context.getResources().getColor(R.color.transmutation));
                break;
        }
        holder.parentView.setOnClickListener(new View.OnClickListener() {
            @Override
            public void onClick(View v) {
                Intent intent = new Intent(context, SpellDetailActivity.class);
                intent.putExtra(SpellDetailActivity.SPELL, spell);
                context.startActivity(intent);
            }
        });
        if(isDatabaseSpell){
            holder.parentView.setOnLongClickListener(new View.OnLongClickListener() {
                @RequiresApi(api = Build.VERSION_CODES.O)
                @Override
                public boolean onLongClick(View v) {
                    int i=0;
                    Boolean alreadyKnowsSpell = false;
                    do{
                        if(character.getSpellsKnown().size() == 0){
                            break;
                        }
                        Spell knownSpell = character.getSpellsKnown().get(i);
                        if(spell.getSpellName().equals(knownSpell.getSpellName())){
                            alreadyKnowsSpell = true;
                            deleteSpell(spell);
                            break;
                        }
                        i++;
                    } while(i<character.getSpellsKnown().size());
                    if(!alreadyKnowsSpell){
                        if(spell.getLevel() == 0){
                            if(cantripsList.size() >= cantripCount){
                                Toast.makeText(context, "You are at your max cantrip count. You cannot add any more cantrips", Toast.LENGTH_SHORT).show();
                            }else{
                                addSpell(spell);
                            }
                        } else if(spell.getLevel() > 0){
                            if(spellsList.size() >= spellCount){
                                Toast.makeText(context, "You are at your max spell count. You cannot add any more spells", Toast.LENGTH_SHORT).show();
                            }else{
                                addSpell(spell);
                            }
                        }
                    }
                    return true;
                }
            });
        }else{
            holder.parentView.setOnLongClickListener(new View.OnLongClickListener() {
                @Override
                public boolean onLongClick(View v) {
                    AlertDialog.Builder adb=new AlertDialog.Builder(context);
                    adb.setTitle("Delete?");
                    adb.setMessage("Are you sure you want to delete " + spell.getSpellName());
                    final int positionToRemove = position;
                    adb.setNegativeButton("Cancel", null);
                    adb.setPositiveButton("Ok", new AlertDialog.OnClickListener() {
                        public void onClick(DialogInterface dialog, int which) {
                            character.getSpellsKnown().remove(spell);
                            spells.remove(positionToRemove);
                            notifyDataSetChanged();
                        }});
                    adb.show();
                    return true;
                }
            });
        }
    }

    @Override
    public int getItemCount() {
        return spells.size();
    }

    public void setSpells(List<Spell> spells) {
        this.spells = spells;
        notifyDataSetChanged();
    }

    public class SpellViewHolder extends RecyclerView.ViewHolder{
        private TextView spellNameTextView;
        private TextView spellLevelTextView;
        private View parentView;

        public SpellViewHolder(@NonNull View view){
            super(view);
            this.parentView = view;
            this.spellNameTextView = view.findViewById(R.id.spell_name);
            this.spellLevelTextView = view.findViewById(R.id.spell_level);
        }
    }

    private void addSpell(Spell spell){
        AlertDialog.Builder adb=new AlertDialog.Builder(context);
        adb.setTitle("Add " + spell.getSpellName());
        adb.setMessage(String.format("Are you sure you want to add %s to your prepared spells?", spell.getSpellName()));
        adb.setNegativeButton("Cancel", null);
        adb.setPositiveButton("Ok", new AlertDialog.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                character.getSpellsKnown().add(spell);
                if(spell.getLevel() == 0){
                    cantripsList.add(spell);
                }else{
                    spellsList.add(spell);
                }
                notifyDataSetChanged();
                Toast.makeText(context, String.format("%s added to spell list", spell.getSpellName()), Toast.LENGTH_SHORT).show();
            }});
        adb.show();
    }

    private void deleteSpell(Spell spell){
        AlertDialog.Builder adb=new AlertDialog.Builder(context);
        adb.setTitle("Delete?");
        adb.setMessage(String.format("You already know %s. Do you want to delete it?", spell.getSpellName()));
        adb.setNegativeButton("Cancel", null);
        adb.setPositiveButton("Ok", new AlertDialog.OnClickListener() {
            public void onClick(DialogInterface dialog, int which) {
                character.getSpellsKnown().remove(spell);
                if(spell.getLevel() == 0){
                    cantripsList.remove(spell);
                }else{
                    spellsList.remove(spell);
                }
                notifyDataSetChanged();
                Toast.makeText(context, String.format("%s removed from spell list", spell.getSpellName()), Toast.LENGTH_SHORT).show();
            }});
        adb.show();
    }
}

File: DDCharacterCreator/app/src/main/java/com/example/ddcharactercreator/SpellAdapter.kt
package com.example.ddcharactercreator

import android.app.AlertDialog
import android.content.Context
import android.content.Intent
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.widget.TextView
import android.widget.Toast
import androidx.recyclerview.widget.RecyclerView
import com.example.ddcharactercreator.SpellAdapter.SpellViewHolder
import java.util.*

class SpellAdapter(private val context: Context, spells: List<Spell>, isDatabaseSpell: Boolean) : RecyclerView.Adapter<SpellViewHolder>() {
    private var spells: List<Spell> = ArrayList()
    private val isDatabaseSpell: Boolean
    private val character: Character = (this as DetailActivity?)!!.character
    private val cantripsList = ArrayList<Spell>()
    private val spellsList = ArrayList<Spell>()
    override fun onCreateViewHolder(parent: ViewGroup, viewType: Int): SpellViewHolder {
        return SpellViewHolder(LayoutInflater.from(context).inflate(R.layout.spell_list_item, parent, false))
    }

    override fun onBindViewHolder(holder: SpellViewHolder, position: Int) {
        for (i in character.getSpellsKnown().indices) {
            val thisSpell = character.getSpellsKnown()[i]
            if (thisSpell.level == 0 && !cantripsList.contains(thisSpell)) {
                cantripsList.add(thisSpell)
            } else if (!cantripsList.contains(thisSpell) && !spellsList.contains(thisSpell)) {
                spellsList.add(thisSpell)
            }
        }
        val spell = spells[position]
        holder.spellNameTextView.text = spell.spellName
        if (spell.level > 0) {
            holder.spellLevelTextView.text = java.lang.String.valueOf(spell.level)
        } else {
            holder.spellLevelTextView.visibility = View.GONE
        }
        when (spell.school) {
            "Abjuration" -> holder.spellNameTextView.setTextColor(context.resources.getColor(R.color.abjuration))
            "Conjuration" -> holder.spellNameTextView.setTextColor(context.resources.getColor(R.color.conjuration))
            "Divination" -> holder.spellNameTextView.setTextColor(context.resources.getColor(R.color.divination))
            "Enchantment" -> holder.spellNameTextView.setTextColor(context.resources.getColor(R.color.enchantment))
            "Evocation" -> holder.spellNameTextView.setTextColor(context.resources.getColor(R.color.evocation))
            "Illusion" -> holder.spellNameTextView.setTextColor(context.resources.getColor(R.color.illusion))
            "Necromancy" -> holder.spellNameTextView.setTextColor(context.resources.getColor(R.color.necromancy))
            "Transmutation" -> holder.spellNameTextView.setTextColor(context.resources.getColor(R.color.transmutation))
        }
        holder.parentView.setOnClickListener {
            val intent = Intent(context, SpellDetailActivity::class.java)
            intent.putExtra(SpellDetailActivity.SPELL, spell)
            context.startActivity(intent)
        }
        if (isDatabaseSpell) {
            holder.parentView.setOnLongClickListener {
                var i = 0
                var alreadyKnowsSpell = false
                do {
                    if (character.getSpellsKnown().size == 0) {
                        break
                    }
                    val knownSpell = character.getSpellsKnown()[i]
                    if (spell.spellName == knownSpell.spellName) {
                        alreadyKnowsSpell = true
                        deleteSpell(spell)
                        break
                    }
                    i++
                } while (i < character.getSpellsKnown().size)
                if (!alreadyKnowsSpell) {
                    if (spell.level == 0) {
                        if (cantripsList.size >= (this as DetailActivity?)!!.cantripCount) {
                            Toast.makeText(context, "You are at your max cantrip count. You cannot add any more cantrips", Toast.LENGTH_SHORT).show()
                        } else {
                            addSpell(spell)
                        }
                    } else if (spell.level > 0) {
                        if (spellsList.size >= (this as DetailActivity?)!!.spellCount) {
                            Toast.makeText(context, "You are at your max spell count. You cannot add any more spells", Toast.LENGTH_SHORT).show()
                        } else {
                            addSpell(spell)
                        }
                    }
                }
                true
            }
        } else {
            holder.parentView.setOnLongClickListener {
                val adb = AlertDialog.Builder(context)
                adb.setTitle("Delete?")
                adb.setMessage("Are you sure you want to delete " + spell.spellName)
                adb.setNegativeButton("Cancel", null)
                adb.setPositiveButton("Ok") { _, _ ->
                    character.getSpellsKnown().remove(spell)
                    spells.drop(position)
                    notifyDataSetChanged()
                }
                adb.show()
                true
            }
        }
    }

    override fun getItemCount(): Int {
        return spells.size
    }

    fun setSpells(spells: List<Spell>) {
        this.spells = spells
        notifyDataSetChanged()
    }

    inner class SpellViewHolder(val parentView: View) : RecyclerView.ViewHolder(parentView) {
        val spellNameTextView: TextView = parentView.findViewById(R.id.spell_name)
        val spellLevelTextView: TextView = parentView.findViewById(R.id.spell_level)

    }

    private fun addSpell(spell: Spell) {
        val adb = AlertDialog.Builder(context)
        adb.setTitle("Add " + spell.spellName)
        adb.setMessage(String.format("Are you sure you want to add %s to your prepared spells?", spell.spellName))
        adb.setNegativeButton("Cancel", null)
        adb.setPositiveButton("Ok") { _, _ ->
            character.getSpellsKnown().add(spell)
            if (spell.level == 0) {
                cantripsList.add(spell)
            } else {
                spellsList.add(spell)
            }
            notifyDataSetChanged()
            Toast.makeText(context, String.format("%s added to spell list", spell.spellName), Toast.LENGTH_SHORT).show()
        }
        adb.show()
    }

    private fun deleteSpell(spell: Spell) {
        val adb = AlertDialog.Builder(context)
        adb.setTitle("Delete?")
        adb.setMessage(String.format("You already know %s. Do you want to delete it?", spell.spellName))
        adb.setNegativeButton("Cancel", null)
        adb.setPositiveButton("Ok") { _, _ ->
            character.getSpellsKnown().remove(spell)
            if (spell.level == 0) {
                cantripsList.remove(spell)
            } else {
                spellsList.remove(spell)
            }
            notifyDataSetChanged()
            Toast.makeText(context, String.format("%s removed from spell list", spell.spellName), Toast.LENGTH_SHORT).show()
        }
        adb.show()
    }

    init {
        this.spells = spells
        this.isDatabaseSpell = isDatabaseSpell
    }
}